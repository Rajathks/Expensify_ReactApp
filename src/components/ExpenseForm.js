import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

import "react-dates/initialize";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      notes: props.expense ? props.expense.notes : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createDate: props.expense ? moment(props.expense.createDate) : moment(),
      CalenderFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return {
        description,
      };
    });
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => {
      return {
        notes: note,
      };
    });
    //console.log("Note Call",note);
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return {
          amount: amount,
        };
      });
    }
  };

  onDateChange = (createDate) => {
    if (createDate) {
      this.setState(() => ({ createDate }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ CalenderFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => {
        return {
          error: "please add description and amount values",
        };
      });
    } else {
      this.setState(() => {
        return {
          error: "",
        };
      });
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createDate: this.state.createDate.valueOf(),
        notes: this.state.notes,
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className ="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Add Description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />

        <input
          type="text"
          placeholder="Add Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createDate}
          onDateChange={this.onDateChange}
          focused={this.state.CalenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add Optional Note"
          className="textarea"
          value={this.state.notes}
          onChange={this.onNoteChange}
        ></textarea>
        <div>
        <button className="button">Add Expense</button>
        </div>
       
      </form>
    );
  }
}

export default ExpenseForm;
