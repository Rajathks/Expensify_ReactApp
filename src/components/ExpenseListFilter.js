import React from "react";
import { connect } from "react-redux";
import {SearchText,SortByDate,SortByAmount,StartDateChanger,EndDateChanger} from "../actions/filters";
import { DateRangePicker } from "react-dates";
import 'react-dates/initialize';

export class ExpenseListFilter extends React.Component {
  state = {
    calenderFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
     // console.log("Here");
      
    this.props.StartDateChanger(startDate);
    this.props.EndDateChanger(endDate);
  };

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = (e) => {
    this.props.SearchText(e.target.value);
  }

  onSortChange = (e) => {
    e.target.value === "amount"
      ? this.props.SortByAmount()
      : this.props.SortByDate();
  }



  render() {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        <select onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
      
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          startDateId="start"
          endDateId="end"
          showClearDates={true}
        />

    
        
      </div>
    );
  }
}
const mapDispatchtoProps = (dispatch) => {
    return {
        StartDateChanger : (startDate) => dispatch(StartDateChanger(startDate)),
        EndDateChanger : (endDate) => dispatch(EndDateChanger(endDate)),
        SearchText : (text) => dispatch(SearchText(text)),
        SortByAmount : () => dispatch(SortByAmount()),
        SortByDate : () => dispatch(SortByDate()),
    }
}

const mapStatetofilter = (state) => {
  return {
    filters: state.filters,
  };
};
export default connect(mapStatetofilter,mapDispatchtoProps)(ExpenseListFilter);
