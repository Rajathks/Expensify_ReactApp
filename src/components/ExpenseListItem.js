import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ description, id, amount, createDate }) => (
  <Link className="list_item" to={`/edit/${id}`}>
    <div >
     <h3 className="list_item__title"> {description} </h3> 
      <span className="list_item__subtitle">{moment(createDate).format(" MMMM Do - YYYY")}</span>
    </div>
    <h3 className="list_item__amount">{numeral(amount / 100).format("$0,00,000")}</h3>
  </Link>
);

export default ExpenseListItem;
