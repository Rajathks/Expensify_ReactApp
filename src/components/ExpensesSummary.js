import React from "react";

import getVisibleExpenses from "../selectors/expenses";
import {connect} from "react-redux";
import numeral from "numeral";
import ExpensesTotal from "../selectors/Expenses-total";

export const ExpensesSummary = ({expenseCount , expenseTotal}) => {
    const expenseWord = expenseCount === 1 ? "expense" : "expenses";
    const fExpense = numeral(expenseTotal/100).format("$0,00,000");
    return (
        <div>
            <p> There are a total of {expenseCount} {expenseWord} and the the total expenditure is : {fExpense}  </p>
        </div>
    );
}

const mapStatetoProps = (state) => {
const visibleExpense = getVisibleExpenses(state.expenses,state.filters);
return {
    expenseCount : visibleExpense.length,
    expenseTotal : ExpensesTotal(visibleExpense)
}
}

export default connect(mapStatetoProps)(ExpensesSummary);