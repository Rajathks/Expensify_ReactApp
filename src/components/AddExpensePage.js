import React from "react";
import ExpenseForm from "./ExpenseForm";
import { AddExpenses } from "../actions/expenses";
import { connect } from "react-redux";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.AddExpenses(expense);
    this.props.history.push("/");
  };


  render() {
      return (
        <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
      );
   
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    AddExpenses: (expense) => dispatch(AddExpenses(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
