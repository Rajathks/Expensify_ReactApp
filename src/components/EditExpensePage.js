import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { EditExpense } from "../actions/expenses";
import { RemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.EditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onClick = () => {
    this.props.RemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />

        <button onClick={this.onClick}>Remove Expense</button>
      </div>
    );
  }
}

const mapDispatchtoProps = (dispatch,props) => {
    return {
        
        EditExpense: (id,expense) => dispatch(EditExpense(id,expense)),
        RemoveExpense : (data) => dispatch(RemoveExpense(data))
    };
}

const mapStatetoProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id;
    }),
  };
};

export default connect(mapStatetoProps,mapDispatchtoProps)(EditExpensePage);
