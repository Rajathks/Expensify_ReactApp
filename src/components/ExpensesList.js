import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import GetVisibleExpenses from "../selectors/expenses";
import getVisibleExpenses from "../selectors/expenses";

export const ExpensesList = (props) => {
  return (
    <div>
      <h2> Here is your Expense List</h2>

      {props.expenses.map((expense) => {
        return <ExpenseListItem {...expense} key={expense.id}/>;
      })}
     
    </div>
  );
};

const mapStatetoProp = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses,state.filters)
  };
};

export default connect(mapStatetoProp)(ExpensesList);
