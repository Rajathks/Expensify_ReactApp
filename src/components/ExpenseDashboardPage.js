import React from "react";
import ExpensesList from "./ExpensesList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpensesSummary from "../components/ExpensesSummary";


const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilter/>
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;