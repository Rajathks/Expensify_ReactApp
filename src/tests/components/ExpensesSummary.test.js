import {shallow} from "enzyme";
import React from "react";
import {ExpensesSummary} from "../../components/ExpensesSummary"; 

test("Should render expenses Summary page with one expense" , () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={235} />);
    expect(wrapper).toMatchSnapshot();


});


test("Should render expenses Summary page with multiple expense" , () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={23} expenseTotal={29090935} />);
    expect(wrapper).toMatchSnapshot();

});