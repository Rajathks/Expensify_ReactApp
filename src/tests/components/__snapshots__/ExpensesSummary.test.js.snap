// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expenses Summary page with multiple expense 1`] = `
<div>
  <p>
     There are a total of 
    23
     
    expenses
     and the the total expenditure is : 
    $290,909
      
  </p>
</div>
`;

exports[`Should render expenses Summary page with one expense 1`] = `
<div>
  <p>
     There are a total of 
    1
     
    expense
     and the the total expenditure is : 
    $2
      
  </p>
</div>
`;
